### PR TITLE
Logic to perform deployment to Cloudhub V2

### LINKED ISSUE #46: CloudHub v2 Depoyment Support
Enhance the deploy library and maven plugin to support deploying to CloudHub v2.  The Mule Maven Plugin added support for this in release 3.7.0 and Cloudhub v2 is now generally available.

### PR DESCRIPTION
**Pull request by gteixeira1**
Resolves #46

### CODE REVIEW

**Comment by manikmagar** [comment]
@gteixeira1 @adesjardin  [CloudHub 2.0 Deployment Model](https://docs.mulesoft.com/mule-runtime/4.4/deploy-to-cloudhub-2#cloudhub2-deploy-reference) seems to have a lot of overlap with [Runtime Fabric Deployment Model](https://docs.mulesoft.com/mule-runtime/4.4/deploy-to-rtf#rtf-deploy-reference). Instead of making this new module as CH 2 specific, should we abstract this model so we could support RTF deployment with any minor extensions to this base mode?

File: library/src/main/java/com/avioconsulting/mule/deployment/api/models/CloudhubV2DeploymentRequest.groovy [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+import com.fasterxml.jackson.databind.ObjectMapper
+import groovy.json.JsonOutput
+import groovy.transform.ToString
+import org.apache.http.HttpEntity
+import org.apache.http.entity.ContentType
+import org.apache.http.entity.mime.HttpMultipartMode
+import org.apache.http.entity.mime.MultipartEntityBuilder
+
+@ToString
+class CloudhubV2DeploymentRequest extends FileBasedAppDeploymentRequest {
```
Comment (manikmagar) [inline library/src/main/java/com/avioconsulting/mule/deployment/api/models/CloudhubV2DeploymentRequest.groovy:13]: @gteixeira1 Are we not supposed to reuse the config between RTF and CH v2?

Comment (gteixeira1) [reply library/src/main/java/com/avioconsulting/mule/deployment/api/models/CloudhubV2DeploymentRequest.groovy:13]: I will perform this adjustment. Thanks

File: library/src/main/java/com/avioconsulting/mule/deployment/api/models/CloudhubV2DeploymentRequest.groovy [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+                                enforceDeployingReplicasAcrossNodes: workerSpecRequest.replicasAcrossNodes,
+                                forwardSslSession: workerSpecRequest.forwardSslSession,
+                                disableAmLogForwarding: workerSpecRequest.disableAmLogForwarding,
+                                generateDefaultPublicUrl: workerSpecRequest.publicURL
+                        ],
+                        replicas: workerSpecRequest.workerCount
+                ]
+        ] as Map<String, String>
+        result
+    }
```
Comment (manikmagar) [inline library/src/main/java/com/avioconsulting/mule/deployment/api/models/CloudhubV2DeploymentRequest.groovy:64]: @gteixeira1 this code is the same as in the parent class. Is there a reason we need it here as well?

File: library/src/main/java/com/avioconsulting/mule/deployment/api/models/CloudhubV2DeploymentRequest.groovy [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+    String getCloudhubAppInfoAsJson() {
+        JsonOutput.toJson(this.cloudhubAppInfo)
+    }
```
Comment (manikmagar) [inline library/src/main/java/com/avioconsulting/mule/deployment/api/models/CloudhubV2DeploymentRequest.groovy:68]: This method is also the same as in parent. Do we need to redefine it here? 

Also, should it be named something non-ch specific? Eg. `getAppInfoAsJson()`.

File: library/src/main/java/com/avioconsulting/mule/deployment/dsl/CloudhubV2Context.groovy [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    CloudhubV2DeploymentRequest createDeploymentRequest() {
+        def errors = findErrors()
+        def specs = workerSpecs
+        errors += specs.findErrors('workerSpecs')
+        def autoDiscovery = this.autoDiscovery
+        errors += autoDiscovery.findErrors('autoDiscovery')
+        if (errors.any()) {
+            def errorList = errors.join('\n')
+            throw new Exception("Your deployment request is not complete. The following errors exist:\n${errorList}")
+        }
```
Comment (manikmagar) [inline library/src/main/java/com/avioconsulting/mule/deployment/dsl/CloudhubV2Context.groovy:16]: Maybe extract to a separate method - validateContext() in the parent. Code is duplicated.

File: library/src/main/java/com/avioconsulting/mule/deployment/dsl/MuleDeployContext.groovy [hunk 5]
```diff
@@ -10,12 +10,14 @@ class MuleDeployContext extends BaseContext {
...
     private CloudhubContext cloudHubApplication = new CloudhubContext()
+    private CloudhubV2Context cloudHubV2Application = new CloudhubV2Context()
+    private RuntimeFabricContext runtimeFabricApplication = new RuntimeFabricContext()
     private OnPremContext onPremApplication = new OnPremContext()
     private FeaturesContext enabledFeatures = new FeaturesContext()
 
     def findErrors() {
         List<String> errors = super.findErrors()
-        if (!cloudHubSet && !onPremSet) {
+        if (!cloudHubSet && !cloudHubV2Set && !onPremSet) {
```
Comment (manikmagar) [inline library/src/main/java/com/avioconsulting/mule/deployment/dsl/MuleDeployContext.groovy:20]: Should it also include `runtimeFabricSet` flag check?

File: library/src/main/java/com/avioconsulting/mule/deployment/dsl/MuleDeployContext.groovy [hunk 6]
```diff
@@ -51,7 +61,7 @@ class MuleDeployContext extends BaseContext {
             def errorList = errors.join('\n')
             throw new Exception("Your file is not complete. The following errors exist:\n${errorList}")
         }
-        if (onPremSet && cloudHubSet) {
+        if (onPremSet && (cloudHubSet || cloudHubV2Set)) {
```
Comment (manikmagar) [inline library/src/main/java/com/avioconsulting/mule/deployment/dsl/MuleDeployContext.groovy:64]: Missing check for RTF?

File: library/src/main/java/com/avioconsulting/mule/deployment/api/models/CloudhubV2DeploymentRequest.groovy (hunk 2 above)
Comment (gteixeira1) [reply library/src/main/java/com/avioconsulting/mule/deployment/api/models/CloudhubV2DeploymentRequest.groovy:64]: The difference is CH v2 requires "vCores" information on the body, which is not used in RTF.
I've refactored the class to use the method from parent and just add additional vCore information

File: library/src/main/java/com/avioconsulting/mule/deployment/api/models/CloudhubV2DeploymentRequest.groovy (hunk 3 above)
Comment (gteixeira1) [reply library/src/main/java/com/avioconsulting/mule/deployment/api/models/CloudhubV2DeploymentRequest.groovy:68]: Refactored the previous method, which will be used by this method

File: library/src/main/java/com/avioconsulting/mule/deployment/dsl/CloudhubV2Context.groovy (hunk 4 above)
Comment (gteixeira1) [reply library/src/main/java/com/avioconsulting/mule/deployment/dsl/CloudhubV2Context.groovy:16]: Code refactored and method on parent class added

File: library/src/main/java/com/avioconsulting/mule/deployment/dsl/MuleDeployContext.groovy (hunk 6 above)
Comment (gteixeira1) [reply library/src/main/java/com/avioconsulting/mule/deployment/dsl/MuleDeployContext.groovy]: Validation was added and also test was improved to validate RTC and CH2

File: library/src/main/java/com/avioconsulting/mule/deployment/dsl/MuleDeployContext.groovy (hunk 5 above)
Comment (gteixeira1) [reply library/src/main/java/com/avioconsulting/mule/deployment/dsl/MuleDeployContext.groovy]: Validation was added and also test was improved to validate RTC and CH2

File: maven-plugin/examples/mule4_project/muleDeploy_ch2.groovy [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+            scopeLoggingConfiguration {
+                scope 'com.pkg.debug'
+                logLevel 'DEBUG'
+            }
+        }
+        target 'Cloudhub-US-West-2'
+        provider 'MC'
+        environment 'DEV'
+        replicas '1'
+        vCores '0.1'
```
Comment (manikmagar) [inline maven-plugin/examples/mule4_project/muleDeploy_ch2.groovy:51]: @gteixeira1  Why are worker specs inline with the application context? Shouldn't it be under workerSpecs {} like CloudHub v1 - https://github.com/avioconsulting/mule-deploy-library/blob/master/library/examples/allOptions.groovy#L86? I think we should have a consistent structure if possible.

File: library/src/main/java/com/avioconsulting/mule/deployment/dsl/WorkerV2SpecContext.groovy [hunk 8]
```diff
@@ -0,0 +1,35 @@
+package com.avioconsulting.mule.deployment.dsl
+
+import com.avioconsulting.mule.deployment.api.models.WorkerSpecRequest
+import com.avioconsulting.mule.deployment.api.models.UpdateStrategy
+import com.avioconsulting.mule.deployment.api.models.VCoresSize
+
+class WorkerV2SpecContext extends BaseContext {
```
Comment (manikmagar) [inline library/src/main/java/com/avioconsulting/mule/deployment/dsl/WorkerV2SpecContext.groovy:7]: Is this for `models/WorkerSpecRequest.groovy`? Can name be consistent if that is the case?

File: maven-plugin/examples/mule4_project/muleDeploy_ch2.groovy (hunk 7 above)
Comment (gteixeira1) [reply maven-plugin/examples/mule4_project/muleDeploy_ch2.groovy:51]: Changes made and properties moved into workerSpecs

**Review by adesjardin — DISMISSED** [review]

File: library/src/main/java/com/avioconsulting/mule/deployment/api/models/deployment/RuntimeFabricDeploymentRequest.groovy [hunk 9]
```diff
@@ -1,13 +1,14 @@
...
-
 import com.avioconsulting.mule.deployment.api.models.WorkerSpecRequest
 import com.avioconsulting.mule.deployment.internal.models.CloudhubAppProperties
 import groovy.json.JsonOutput
 import groovy.transform.ToString
 
 @ToString
 class RuntimeFabricDeploymentRequest extends ExchangeAppDeploymentRequest {
+
+    final MAX_SIZE_APPLICATION_NAME = 42
```
[the remark this answers is not in the document]
Comment (manikmagar) [reply library/src/main/java/com/avioconsulting/mule/deployment/api/models/deployment/RuntimeFabricDeploymentRequest.groovy:12]: Should be on [CH v1](https://help.mulesoft.com/s/article/CloudHub-Application-Name-is-Limited-to-42-Characters) too. For standalone, the app-limit may depend on the MULE_HOME path due to overall File PATH restrictions from OS. Having it configurable for standalone might help to be consistent in naming strategy.

File: library/examples/muleDeploy_ch2_sharedspace_full.groovy.groovy [hunk 10]
```diff
@@ -0,0 +1,23 @@
+muleDeploy {
+    cloudHubV2Application {
+        environment params.env
+        applicationName
+        applicationName {
```
Comment (manikmagar) [inline library/examples/muleDeploy_ch2_sharedspace_full.groovy.groovy:5]: Looks like a left duplicate by mistake.

File: library/src/main/java/com/avioconsulting/mule/deployment/api/models/deployment/ApplicationName.groovy [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+            } else {
+                normalizedAppName = buildAppNameByOptions().toLowerCase()
+            }
+        }
+        runNameLengthGuard(normalizedAppName)
+        return normalizedAppName
+    }
+
+    private void runGuards(){
+        assert (baseAppName != null && !baseAppName.isBlank() && !baseAppName.contains(" ")) : "you should specify an non-empty baseAppName. It shouldn't contain spaces as well"
```
Comment (manikmagar) [inline library/src/main/java/com/avioconsulting/mule/deployment/api/models/deployment/ApplicationName.groovy:41]: 1. Are special characters other than spaces allowed as app name? Eg. `some#app*&$name`. Current validation accepts that name. Maybe just restricting it to Alphanumeric and Dash is better?
2. Error message formatting, could be a simple message `Application name cannot be empty and cannot contain spaces.` If more characters are not allowed then probably we can just say that "Name must be alphanumeric and can include dash."

File: library/src/main/resources/evaluate.gdsl.vm [hunk 12]
```diff
@@ -211,4 +210,13 @@ contributor([closureScopeContext]) {
         #stringParam('clientId')
         #stringParam('clientSecret')
     }
+    call = enclosingCall('applicationName')
+    if (call) {
+        #stringParam('baseAppName')
+        #boolParam('usePrefix')
+        #boolParam('useSuffix')
+        #stringParam('prefix')
+        #stringParam('suffix')
```
Comment (manikmagar) [inline library/src/main/resources/evaluate.gdsl.vm:219]: Curious to know if we need pairs of properties eg usePrefix and prefix? Can the existence of valid `prefix` or `suffix` strings not be used to decide and reduce the configuration elements?

File: maven-plugin/pom.xml [hunk 13]
```diff
@@ -110,13 +110,19 @@
...
             <scope>test</scope>
         </dependency>
         <dependency>
-            <groupId>junit</groupId>
-            <artifactId>junit</artifactId>
-            <version>4.13.1</version>
+            <groupId>org.junit.platform</groupId>
+            <artifactId>junit-platform-suite-engine</artifactId>
+            <version>${junit.platform.version}</version>
+            <scope>test</scope>
```
Comment (manikmagar) [inline maven-plugin/pom.xml:120]: Should we move this dependencyManagement in parent pom?

**Pull request merged**